### ISSUE TITLE
Dual GPU cuda core dumped.

### ISSUE THREAD

**Issue by aldakata** [open-mmlab/mmdetection3d#1055]
I am on a dual GPU system with pytorch 1.10 and cuda 1.13
When I run inference_detector on centerpoint with device=cuda:1 it exits upon this exception.
I have no issue when running the same code on cuda:0.

Both GPU are the same model with the same drivers installed.

**Checklist**
1. I have searched related issues but cannot get the expected help.
2. The bug has not been fixed in the latest version.

**Describe the bug**
Program exits after error. Traceback provided below.

**Reproduction**
1. What command or script did you run?
I run 

```python
    # build the model from a config file and a checkpoint 
    CONFIG = '/mmdetection3d/configs/centerpoint/centerpoint_01voxel_second_secfpn_circlenms_4x8_cyclic_20e_nus.py'
    CHECKPOINT = '/mmdetection3d/checkpoints/centerpoint_01voxel_second_secfpn_circlenms_4x8_cyclic_20e_nus_20201001_135205-5db91e00.pth'
    model = init_model(CONFIG, CHECKPOINT, device='cuda:1')
    pointcloud = '/my/pointcloud.bin'
    result, data = inference_detector(self.model, pointcloud)
```

2. Did you make any modifications on the code or config? Did you understand what you have modified?
I didn't.
3. What dataset did you use?
Custom dataset.
**Environment**

1. Please run `python mmdet3d/utils/collect_env.py` to collect necessary environment infomation and paste it here.
2. You may add addition that may be helpful for locating the problem, such as
    - How you installed PyTorch [e.g., pip, conda, source]
    - Other environment variables that may be related (such as `$PATH`, `$LD_LIBRARY_PATH`, `$PYTHONPATH`, etc.)

**Error traceback**
If applicable, paste the error trackback here.
```bash
terminate called after throwing an instance of 'c10::CUDAError'
  what():  CUDA error: an illegal memory access was encountered
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Exception raised from create_event_internal at /opt/conda/conda-bld/pytorch_1634272168290/work/c10/cuda/CUDACachingAllocator.cpp:1211 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x42 (0x7f52d2b97d62 in /home/cavride/anaconda3/envs/mmlab/lib/python3.7/site-packages/torch/lib/libc10.so)
frame #1: <unknown function> + 0x1c613 (0x7f52d2dfa613 in /home/cavride/anaconda3/envs/mmlab/lib/python3.7/site-packages/torch/lib/libc10_cuda.so)
frame #2: c10::cuda::CUDACachingAllocator::raw_delete(void*) + 0x1a2 (0x7f52d2dfb022 in /home/cavride/anaconda3/envs/mmlab/lib/python3.7/site-packages/torch/lib/libc10_cuda.so)
frame #3: c10::TensorImpl::release_resources() + 0xa4 (0x7f52d2b81314 in /home/cavride/anaconda3/envs/mmlab/lib/python3.7/site-packages/torch/lib/libc10.so)
frame #4: <unknown function> + 0x294dd9 (0x7f5257750dd9 in /home/cavride/anaconda3/envs/mmlab/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #5: <unknown function> + 0xae2f59 (0x7f5257f9ef59 in /home/cavride/anaconda3/envs/mmlab/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #6: THPVariable_subclass_dealloc(_object*) + 0x2b9 (0x7f5257f9f279 in /home/cavride/anaconda3/envs/mmlab/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
<omitting python frames>
frame #31: __libc_start_main + 0xe7 (0x7f52e8255bf7 in /lib/x86_64-linux-gnu/libc.so.6)
Aborted (core dumped)
```

**Bug fix**
I checked what was the current device in init_model after model.to(device) with:
```python
    print(f'HELLO I AM MMDETECTION3D {torch.cuda.current_device()}')
````
and it always printed 0 regardless of the device I entered.

I fixed my issue adding this line right before model.to(device)  of init_model() in inference.py :
```python
    torch.cuda.set_device(device)
    model.to(device)
```
If I print the torch.cuda.current_device() after the fix it now is 0 or 1 accordingly! :)

**Comment by Tai-Wang** [comment]
Thanks for your contribution! We have just fixed a similar issue in PR #1045, which is also a general problem in other codebases (such as mmdetection). A little question is that what data is loaded onto an incorrect device, data or model? From my understanding, the model should be converted to the specified device via model.to(device) and data is also converted in `inference_detector`?

**Comment by aldakata** [comment]
I don't know. I think the expected behaviour is as you say. I will have access to the PC on Monday, I will do some more debugging to understand what is happening.

It is clear that ```model.to(device) ``` is doing something since with ```device='cuda:0'``` it works as expected and ```device='cuda:1'``` throws this nasty error.

To be clear, the model loads accordingly, it is at the _inference_detector_ call when the error is thrown.

What I don't know for sure is if after ```model.to(desired_device)``` the ```torch.cuda.current_device()``` is set to _desired_device_ or if it just allocates the tensor _model_ to _desired_device_ without setting it to current device...

If anyone knows the answer to this, or has a multiple GPU system available and can try running this script and say what the two prints returned:
```python
import troch
tensor = torch.randn(2,2)
print(torch.cuda.current_device()) #Print 1
tensor.to('cuda:1') #This device should be different from the current device so the test is meaningful.
print(torch.cuda.current_device()) #Print 2
```
